framework/source/class/qx/ui/tree/core/OpenCloseController.js

```diff
@@ -0,0 +1,152 @@
+{
+  extend: qx.core.Object,
+  
+  /**
+   * @param tree {qx.ui.tree.VirtualTree}
+   *   The tree whose branch open or closed state is to be synchronized to a
+   *   model property.
+   * 
+   * @param model {qx.data.Array}
+   *   The model wherein a property is to be synchronized to the tree
+   *   branches' open or closed states
+   * 
+   * @param openPropertyName {String?"open"}
+   *   The name of the boolean property in the model which controls whether a
+   *   branch in the tree is open or closed. Defaults to "open".
+   */
+  construct: function(tree, model, openPropertyName)
+  {
+    this.base(arguments);
+    
+    // If a property name was specified, use it instead of the default
+    if (openPropertyName)
+    {
+      this.setOpenPropertyName(openPropertyName);
+    }
+    
+    // Save the tree and initialize storage of listener IDs
+    this._tree = tree;
+    this._lids = [];
+    
+    // Sync tree nodes
+    var sync = function(node) {
+      var openPropertyName = this.getOpenPropertyName();
+
+      if (qx.Class.hasProperty(node.constructor, "children")) {
+        node.getChildren().forEach(sync);
+      }
+      
+      if (qx.Class.hasProperty(node.constructor, openPropertyName)) {
+        if (node.get(openPropertyName)) {
+          tree.openNode(node);
+        }
+        else {
+          tree.closeNode(node);
+        }
+      }
+    }.bind(this);
+    sync(model.getItem(0));
+    
+    // Wire change listeners
+    var lid = tree.addListener("open", this._onOpen, this);
+    this._lids.push([tree, lid]);
+    lid = tree.addListener("close", this._onClose, this);
+    this._lids.push([tree, lid]);
+    lid = model.addListener("changeBubble", this._onChangeBubble, this);
+    this._lids.push([model, lid]);
+  },
+  
+  properties :
+  {
+    /**
+     * Name of the property whose value determines whether a branch is open or
+     * not
+     */
+    openPropertyName :
+    {
+      check : "String",
+      nullable : false,
+      init : "open"
+    }
+  },
+  
+  members:
+  {
+    /** The tree which is synced to the model */
+    _tree: null,
+
+    /** Listener IDs that we manage */
+    _lids: null,
+    
+    // event listener for "open" on the tree
+    _onOpen: function(ev)
+    {
+      ev.getData().set(this.getOpenPropertyName(), true);
+    },
+    
+    // event listener for "close" on the tree
+    _onClose: function(ev)
+    {
+      ev.getData().set(this.getOpenPropertyName(), false);
+    },
+    
+    // event listener for model changes
+    _onChangeBubble: function(ev)
+    {
+      var bubble = ev.getData();
+      var modelPropRe;
+
+      // generate a regular expression that identifies model changes that
+      // pertain to the open state of a branch in the tree.
+      modelPropRe = new RegExp("\\." + this.getOpenPropertyName() + "$");
+      
+      // open related? sync it back to the node item.
+      if (modelPropRe.test(bubble.name)) {
```

Comment:
If this is meant to be generic, then relying on change bubbles may be problematic. Using the JSON marshaller, you can disable bubbling for the automatic properties - and then this construct will never notice that the `openPropertyName` changed. I.e. we only enable bubbling when there's absolutely no other way, because it is very ineffective. In one project, we've a large tree. Disabling event propagation makes initializing the model up to 500% faster.

Just a thought: maybe we can use `qx.event.message.Bus` inside the `VirtualTreeItem` objects as a dispatcher and the method I'm currently commenting as the receiver. The `VirtualTreeItem` could establish a listener to the `openPropertyName` of it's model - and when it changes, it puts it's model as the payload on the message bus topic *treeId.open*. Then we've a single event sink and no need for bubbling. And the listener will not be called on every property change.

Drawback: if you want to use this feature, you need a delegate for i.e. *configureItem* in order to configure the treeId and the openPropertyName.

If this code would be included directly to the VirtualTree class (enabled by a property 'syncOpenState'), a tree-name could be omitted.

Reply:
Thanks for the comments, @cajus. I will look into the implementation you suggest. I tried, however, disabling bubbling in the model creation, and this still works. I'm wondering why? See the following gist playground snippet, and notice `false` in the model creation on line 129, which I think should disable change bubbles; and the timer at the end which alters the model: https://gist.github.com/derrell/d5069ab6a981192ac30eff8d969d8565

Reply:
I guess that's because you choose item[0] for the test. It's the one we're listening on.

Reply:
It works even if I use for the test:

```
qx.event.Timer.once(
  function()
  {
    model.getItem(0).getChildren().getItem(0).setOpen(true);
  },
  this,
  2000);
```

But in any case, let me work on the alternate implementation and assuming I can get it working, we can then choose one or the other. Thanks.

Reply:
Sure :+1:. No sure why it works. Either the *true/false* flag for the JSON marshaller is not doing what it says, or I don't know. We've some XML or Protobuf models here, so we don't use the marshaller at all.

Sorry for being a bit offline. The kids... ;-)